### PR TITLE
New INT header and Wireshark plug-in

### PR DESCRIPTION
**Pull request by RandyLevensalor**
#### What does this PR do?
Adds a lua plug-in to wireshark
Documents the proposed INT header
#### Do you have any concerns with this PR?
yes.
#### How can the reviewer verify this PR?
Implement and test the header

#### Questions:
- Have you connected this PR to the issue it resolves?
Issue #57 and #65 
- Does the documentation need an update?
This is a documentation change
- Does this add new dependencies?
No
- Have you added unit or functional tests for this PR?
No

### CODE REVIEW

File: docs/int_header/INT_header.md [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+There are two INT headers that will be used with Transparent Secuirty.  One for the origionating device on the customer premisis and the second is for the network path (Switches and gateways).
+
+## IP Header update
+
+### Encapsulation
+
+We want to preserve the routing and other information on the packet, so we will modify the protocol (IPv4) / next header (IPv6), length and for IPv4 the header checksum.
+
+Define a new L3 protocol.  The protocol for the INT INT header will be 63 "Any local network", since the INT header should never be exposed outside of the service providers network.  There is no specification for portocol 63.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: INT INT -> INT
portocol -> protocol

File: docs/int_header/INT_header.md [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+IPv4:
+
+* Update the total lenght to match the new size.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: lenght -> length

File: docs/int_header/INT_header.md [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+* Set the Protocol to the INT procol number. 253
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: procol -> protocol

File: docs/int_header/INT_header.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+* Update the header checksum
+
+IPv6:
+
+* Set the Next Header to be the identifier for an IPv6 extension header to the INT procol number.  If exisint extension headers are in place, INT will prepend to the first extension header.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: exisint -> existing

File: docs/int_header/INT_header.md [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+* The INT header next header field will be the exisint IPv6 next header value.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: exisint -> existing

File: docs/int_header/INT_header.md [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+* The Hdr Ext Len will be updated appropriatly
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: appropriatly -> appropriately

File: docs/int_header/INT_header.md [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+### Decapsulation
+
+The origional protocol / next header will be restored and the size and checksum will be recalculated as appropriate.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: origional -> original

File: docs/int_header/INT_header.md [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+
+## INT Header
+
+### INT metadata header (64 bits)
+
+This format will be used for both the device INT header and the network path INT header.  The INT instruction bitmask will indicat which data will be used.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: indicat -> indicate

File: docs/int_header/INT_header.md [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+
+* Each INT metadata header is 8B long and should be 0 when initialized
+  * Ver (2b): INT metadata header version. Should be zero for this version
+  * Rep (2b): Replication requested. Should be zero
+  * C (1b): Copy. Should be zero
+  * E (1b): Max Hop Count exceeded. Set to 1 for the device INT header and when the max hop count has been exceeded with the networking header.
+  * Reserved (4b): Always zero.
+* Instruction Count (5b): The number of instructions that are set (i.e., number of
+1’s) in the instruction bitmap. This is 00010 (decimle 2) for the default gateway INT header.  Can be changed to add additional data.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: decimle -> decimal

File: docs/int_header/INT_header.md [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+* Max Hop Count (8b): 1 for the device INT header and configurable for the network header.
+* Total Hop Count (8b): The current hop count for the network header and 1 for the device header.
+* INT instructions are encoded as a bitmap in the 16 bit INT Instruction field: the first 8 bits
+corresponds to a specific standard metadata as specified in Section 3 of the P4 INT spec.  The 9th-11th are defined in the metadata section below.  For the device ID the default is 000000001110000000.  If the originating device is using IPv6, then the originating IP address is not captured and the bit mask for the device INT header will be 000000001100000000.  For the network INT header the bit mask would be 100000000000000000.   Additional metadata can be added for the network INT header for performance and additional use cases.
+  * bit0 (MSB): Switch ID
+  * bit8: Origionating Device MAC (Most signifigant 4 octects)
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: Origionating -> Originating
as with the next 2 lines
octects -> octets (elswhere in file too)

File: docs/int_header/INT_header.md [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+  * bit9: Origionating Device MAC (Least signifigant 2 octects + 2 octects of 0 padding)
+  * bit10: Origionating Device IPv4 address (4 octects)
+* Next header IPv6 / Protocl IPv4 (1 octect)  If this is the last INT (header before the data gram), this will be the initial protocol from the IP header.  If there are subsequent headers then this will be 63 to denote that another INT header is to follow.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: Protocl -> Protocol

File: docs/int_header/INT_header.md [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+* Reservered (1 octects)
+
+### Device INT Metadata (96 bits)
+
+This section is deviating from the INT spec as we are removing the leading bit to indicate the last record.
+
+This metadata will only containe one record and will not be updated on subsequent hops.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: containe -> contain

File: docs/int_header/INT_header.md [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+
+Each metadata record corrosponds to a bit filed in the instruction set and is 4 octects long.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: corrosponds -> corresponds

File: docs/int_header/INT_header.md [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+
+* Origionating Device MAC Most signifigant 4 octects (4 octects)
+* Origionating Device MAC least signifigant 2 octects + 2 octects of padding (4 octects)
+* Origionating Device IPv4 (4 octects)
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: Origionating -> Originating

File: docs/int_header/INT_header.md [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+
+### Network INT Metadata (32 bits per hop)
+
+This section is deviating from the INT spec as we are removing the leading bit to indicate the last record.
+
+This metadat will be updated with each consecutive hop, until the max hop count has been reached.  The additional hop infomration will be inserted between the INT metadata header and the metadata from the previous hop.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: metadat -> metadata

File: docs/int_header/INT_header.md [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+
+Each metadata record corrosponds to a bit filed in the instruction set and is 4 octects long.
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: corrosponds -> corresponds

File: docs/int_header/INT_header.md [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+
+* Switch ID: Unique identfier for the swtich (4 octects)
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: identfier -> identifier

File: docs/int_header/INT_header.md [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+  <td colspan=8 class=xl67>Next Protocol</td>
+  <td colspan=8 class=xl63>Reserved</td>
+ </tr>
+ <tr height=21 style='height:16.0pt'>
+  <td rowspan=3 height=84 class=xl65 style='height:64.0pt'>INT Metadata</td>
+  <td colspan=32 height=21 class=xl67 style='height:16.0pt'>Origionating Device
+  MAC Most signifigant 4 octects<span style='mso-spacerun:yes'> </span></td>
+ </tr>
+ <tr height=21 style='height:16.0pt'>
+  <td colspan=16 height=21 class=xl67 style='height:16.0pt'>Origionating Device
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: Origionating -> Originating

File: docs/int_header/INT_header.md [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+  <td colspan=8 class=xl63>Total Hop Count</td>
+ </tr>
+ <tr height=21 style='height:16.0pt'>
+  <td colspan=16 height=21 class=xl66 style='height:16.0pt'>Instruction Bitmap</td>
+  <td colspan=8 class=xl67>Next Protocol</td>
+  <td colspan=8 class=xl63>Reserved</td>
+ </tr>
+ <tr height=21 style='height:16.0pt'>
+  <td rowspan=3 height=84 class=xl65 style='height:64.0pt'>INT Metadata</td>
+  <td colspan=32 height=21 class=xl67 style='height:16.0pt'>Origionating Device
```
Comment (spisarski) [inline docs/int_header/INT_header.md]: Origionating -> Originating

File: docs/int_header/README.md [hunk 20]
```diff
@@ -0,0 +1,17 @@
+# In-band Network Telemetry Header
+
+## Header definition
+
+The header is defined in [INT Header defintion](INT_header.md)
+
+## Wireshark plug-in
+
+Note: This version of the plug-in matches what is currently implemented in
+the miniment simulator implementation and may be behind the curred documented
```
Comment (spisarski) [inline docs/int_header/README.md]: curred documented -> current documentation

File: docs/int_header/INT_header.md [hunk 21]
```diff
@@ -0,0 +1,320 @@
+# Transparent Secuirty INT header reference definition
+
+The INT header is an L3 wrapper containing in band network telemetry.
+
+These definitions contains modifications and required fieleds for Transpaerent Secuirty.
```
Comment (Sanjhana-Jayagopal) [inline docs/int_header/INT_header.md]: *fields for Transparent Security

File: docs/int_header/INT_header.md [hunk 22]
```diff
@@ -0,0 +1,320 @@
...
+
+  Note: This version of the document is still under discussion and may change.
+
+This header heavily leverages the P4 INT header.  The changes and subset that are required for
+Transparent Security are noted is this document.
```
Comment (Sanjhana-Jayagopal) [inline docs/int_header/INT_header.md]: *noted in this document

File: docs/int_header/INT_header.md [hunk 23]
```diff
@@ -0,0 +1,320 @@
...
+
+The origional header document is located at [INT-current-spec.pdf](https://p4.org/assets/INT-current-spec.pdf)
```
Comment (Sanjhana-Jayagopal) [inline docs/int_header/INT_header.md]: *original

File: docs/int_header/INT_header.md [hunk 24]
```diff
@@ -0,0 +1,320 @@
...
+
+## Overview
+
+The INT header(s) will be inserted between the IP header and the datagram for IPv4 and will be a part of the extended header for IPv6.  The IP header will be updated to indicate that it has an INT header.
+
+The INT header is defined in two portions.  One is the Header for the INT metadata and the second is the actual metadata.
+
+There are two INT headers that will be used with Transparent Secuirty.  One for the originating device on the customer premisis and the second is for the network path (Switches and gateways).
```
Comment (Sanjhana-Jayagopal) [inline docs/int_header/INT_header.md]: *Transparent Security

File: docs/int_header/INT_header.md [hunk 25]
```diff
@@ -0,0 +1,320 @@
...
+
+* Originating Device MAC Most signifigant 4 octets (4 octets)
+* Originating Device MAC least signifigant 2 octets + 2 octets of padding (4 octets)
+* Originating Device IPv4 (4 octets)
+
+### Network INT Metadata (32 bits per hop)
+
+This section is deviating from the INT spec as we are removing the leading bit to indicate the last record.
+
+This metadata will be updated with each consecutive hop, until the max hop count has been reached.  The additional hop infomration will be inserted between the INT metadata header and the metadata from the previous hop.
```
Comment (Sanjhana-Jayagopal) [inline docs/int_header/INT_header.md]: *information

File: docs/int_header/INT_header.md [hunk 26]
```diff
@@ -0,0 +1,320 @@
...
+  <td width=39 style='width:29pt'></td>
+  <td width=39 style='width:29pt'></td>
+  <td width=39 style='width:29pt'></td>
+  <td width=39 style='width:29pt'></td>
+  <td width=39 style='width:29pt'></td>
+  <td width=39 style='width:29pt'></td>
+  <td width=39 style='width:29pt'></td>
+ </tr>
+ <tr height=21 style='height:16.0pt'>
+  <td height=21 style='height:16.0pt'>Octtect</td>
```
Comment (Sanjhana-Jayagopal) [inline docs/int_header/INT_header.md]: *Octet

File: docs/int_header/INT_header.md [hunk 27]
```diff
@@ -0,0 +1,320 @@
...
```
Comment (Sanjhana-Jayagopal) [inline docs/int_header/INT_header.md]: *Octet

**Pull request merged**